### ISSUE TITLE
Avoid subnets that don't have available IP Addresses

### ISSUE THREAD

**Issue by ellistarn** [aws/karpenter#5234]
### Description

**What problem are you trying to solve?**

When a subnet is almost out of IPs, Karpenter will continue to launch nodes in it, leading to the VPC CNI failing to become ready, and the node becoming unready as well. In many cases, there's nothing we can do, but if another subnet has IP addresses, and the workload does not have scheduling constraints that prevent it from running in those zones, we should launch in the subnets with available IPs

One way to approach this would be to minimize the . This would have a side effect of a zonal balance, which c

**How important is this feature to you?**

Managing the ipv4 space is hard, and anything we can do to alleviate these pains would help customers on their path to ipv6.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by martinsmatthews** [comment]
@ellistarn if we have multiple available subnets available for a zone, will karpenter choose the least full subnet like it says in here https://karpenter.sh/v0.32/concepts/nodeclasses/#specsubnetselectorterms ?

**Comment by ellistarn** [comment]
Correct. This is a good point. We don't do this for subnets in different zones
.

**Comment by martinsmatthews** [comment]
Would be really good if we could do this, as this would/could also naturally balance instances across zones which would be a nice HA feature. We have seen that unless there is a topology spread in our deployments, multiple instances get spun up in the same zone.

**Comment by sthapa-ping** [comment]
This is one of the burning issue that we are currently facing with Karpenter, we discussed about this in reinvent. Checking available IPs across the subnets using AWS endpoint and scheduling next instances to the subnet with most IPs seems straight forward. Any approximate release timeline for when we can expect this feature to be added?

**Comment by ellistarn** [comment]
Are you finding that there are 0 remaining IPs and the launch fails, or just a few IPs? Can you provide logs that occur when this happens? Is the failure at the node level? Pod level?

**Comment by martinsmatthews** [comment]
> Are you finding that there are 0 remaining IPs and the launch fails, or just a few IPs?

No we were seeing that without pod anti affiinity to force them to be spread across zones we'd often end up with multiple nodes in the 1 zone and none in the other 2 zones, then all the pods trying to spin up and exhaust the subnet. Note that this is using lots of small pods and relatively small subnets (/26). We weren't seeing node launches fail.

We're not seeing this issue as we moved to larger subnets and added the anti affinity which means the nodes are spread across the zones/subnets more evenly.

Am happy to recreate this and send some logs if that would be helpful @ellistarn ?

**Comment by ellistarn** [comment]
@martinsmatthews , have you completely exhausted the ipv4 space? Is it possible to add another subnet with more IPs? Why are you so constrained? 

I'm working on an idea for EKS networking and I'd love to chat more over slack.

**Comment by martinsmatthews** [comment]
Coming back to the use of `topologySpreadConstraints` to solve this issue and stripe pods over AZs (and thus over subnets) this doesn't work perfectly. I spun up 10 deployments with 3 replicas and with a `topologySpreadConstraints` like (plus a large enough resource ask to end up with only one pod per node) with a nodepool that had 3 subnets defined, 1 in each of 3 AZs:
```
      topologySpreadConstraints:
        - maxSkew: 1
          topologyKey: topology.kubernetes.io/zone
          whenUnsatisfiable: ScheduleAnyway
          labelSelector:
            matchLabels:
              app: topology-test-{{ count }}
```
and this gave me 30 nodes, 10 in each AZ - bingo. But... then I dropped the replica count to 2 and spun up 15 deployments and this gave me a very uneven spread of nodes:
- us-west-2a: 19
- us-west-2b: 7
- us-west-2c: 14

Obviously this is artificial, but this does again highlight a need for an option to balance nodes across AZs even if this is not the default. And not just for subnet usage reasons, there is an HA risk here for when we lose a zone - it doesn't happen often, but when it does and we were skewed like this in a production cluster, it wouldn't be pretty.

**Comment by Shadowssong** [comment]
Has anyone come up with a solution to work around this?  We have twice run into a situation where Karpenter overloads a single AZ (with two /20 subnets) and both subnets run out of IP's.  The skew was quite extreme (200 nodes in 1 AZ, 50 in the other 3) and it seems like a burst of spot requests may have put them all on the same AZ, but it still seems odd that Karpenter doesn't attempt to do any kind of load balancing across the AZ's.  Our current work around is to define a nodepool per AZ and use the cpu/memory limits to faux-limit the number of nodes to prevent IP exhaust, but this results in a lot of nodepools and ec2nodeclass's.  This only works because we currently use a fairly strict set of instance types so we know what their max pod count would be.  If anyone has come up with a better solution please share!

**Comment by snieg** [comment]
> We have twice run into a situation where Karpenter overloads a single AZ (with two /20 subnets) and both subnets run out of IP's. 
Same issue here.

After migrating to karpenter, it started favoring one zone (eu-west-1c), despite having over 1000 addresses available in other subnets, which ended up with us running out of addresses in eu-west-1c.

```
╰─❯ kubectl get nodes -L  worker-type,group,topology.kubernetes.io/zone --sort-by=.metadata.creationTimestamp --no-headers   -l group=default | awk '{print $8}'  |sort | uniq -c
   7 eu-west-1a
   7 eu-west-1b
  25 eu-west-1c
  ```
  
different cluster but still running on cluster-autoscaler:
  
  ```
    18 eu-west-1a
  17 eu-west-1b
  18 eu-west-1c
  ```

**Comment by maxforasteiro** [comment]
hey, can we have some 👀 in the opened PRs? @ellistarn

**Comment by AayushBangroo** [comment]
Just had pods in our production clusters stuck because karpenter keeps launching nodes in a single subnet even when multiple subnets specified. Any idea when the fix will be merged and released?

**Comment by Vacant2333** [comment]
@AayushBangroo Based on the current logic, `karpenter` will choose the subnet with the most predicted margin to create a Node. You can temporarily create a subnet with more margin in a zone with insufficient resources to solve this problem.

**Comment by Vacant2333** [comment]
Of course, I also hope that my PR can be reviewed & merged as soon as possible. We May need more approvers/maintainers, because they are really too busy with maintainers at the moment.
cc @njtran @jonathan-innis

**Comment by AayushBangroo** [comment]
> @AayushBangroo Based on the current logic, `karpenter` will choose the subnet with the most predicted margin to create a Node. You can temporarily create a subnet with more margin in a zone with insufficient resources to solve this problem.

Thanks for the reply. What do you mean by most predicted margin?

**Comment by Vacant2333** [comment]
@AayushBangroo Simply put, for an AZ, karpenter will select one of the Subnets for that AZ, and try to select the Subnet with the most predicted remaining IPs. So you only need to create a *bigger* subnet for the AZ which may pending pods.

**Comment by dfsdevops** [comment]
It seems to me like theres two separate issues:

1. Karpenter doesn't seem to round robin it's AZ selection unless working around scheduling constraints on workloads.
2. Has no way to detect if a subnet is out of IPs.

Seems like folks are talking about fixing item no. 2, but I'm still not sure what explains the behavior of no. 1. Unless I missed something in this thread. It seems to me that both issues should be addressed, one just makes the other a little more apparent but I see them as separate issues.

**Comment by saurav-agarwalla** [comment]
Cross-posting link to @jmdeal's response on the PR: https://github.com/aws/karpenter-provider-aws/pull/7623#issuecomment-2881460627

**Comment by viktor42b** [comment]
What's the link with https://github.com/aws/karpenter-provider-aws/pull/8199?

**Comment by DerekFrank** [comment]
#8199 should help mitigate IP exhaustion if (and only if) the IP exhaustion is limited to a specific AZ. It unfortunately doesn't solve the root of the issue

**Comment by andretapxure** [comment]
The workaround I've adopted is to not run spot instances. In contact with a AWS engineer, he mentioned that the problem occurs because the avoided zone is too expensive so I removed the spot instances and karpenter started to distribute nodes randomly in my four zones

**Comment by bwagner5** [comment]
> The workaround I've adopted is to not run spot instances. In contact with a AWS engineer, he mentioned that the problem occurs because the avoided zone is too expensive so I removed the spot instances and karpenter started to distribute nodes randomly in my four zones

You could also use a zonal topology spread balance pods across the zones to help a bit rather than not using Spot at all. This will result in acquiring capacity within a single zone per request (when a new node is needed).

**Comment by maxforasteiro** [comment]
> > The workaround I've adopted is to not run spot instances. In contact with a AWS engineer, he mentioned that the problem occurs because the avoided zone is too expensive so I removed the spot instances and karpenter started to distribute nodes randomly in my four zones
> 
> You could also use a zonal topology spread balance pods across the zones to help a bit rather than not using Spot at all. This will result in acquiring capacity within a single zone per request (when a new node is needed).

the problem is that most topology spread constraints are set to `ScheduleAnyway`, so the k8s scheduler does not require it to be in a different zone, it just prefers a different zone if it is available. Since Karpenter is optimized for cost performance, any $0,00001 difference between zone A and zone B spot instances will tend to schedule all new pods in the cheapest zone.

We have followed some AWS Karpenter dev recommendation to split our NodePools per Business Unit, and enforce TSC on zones with `DoNotSchedule` so k8s scheduler really forces Karpenter to spread them between zones. This also helps to visualize cost attributions per tag in AWS Cost Explorer (which is a good plus) but it does increase the costs since optimization is limited per business unit.

**Comment by jmdeal** [comment]
Since there aren't any actionable steps for this feature at this time, I'm going to convert this into a discussion. I will copy over my response from #7623 as well since it has the context for why this isn't currently actionable from Karpenter. 

The TLDR is that the core issue is pods can be scheduled to nodes without available IPs. Until that core issue is addressed upstream (likely through coordination between the scheduler and kubelet / CNI) any changes on Karpenter's side will be mitigations that we don't believe improve the current state of the world.
---
I've reviewed with the team, and we do not see a good path forward for this at the moment. I've summarized our rationale and the potential steps forward here:

## Summary of current state

Karpenter combines two sources to estimate the number of IPs available in a given subnet:

* The value from DescribeSubnets
* The minimum number of IPs consumed by the set of instances sent to CreateFleet (upper-bound based on maxPods value for the smallest instance)

When new values are discovered via DescribeSubnets, the tracked in-flight IPs are reset for that subnet. This estimate is not a lower-bound for the number of IPs in the subnet for two reasons:

* Karpenter currently doesn't update the in-flight IPs based on the resolved instance from CreateFleet (the resolved maxPods value may be higher).
* When IP values are reset based on the DescribeSubnets results, this discards any information about the maximum IP consumption by launched instances and only reflects the currently consumed IPs

Currently this information is only used to prioritize a subnet within a given zone, so it not being a strict lower-bound isn't problematic. However, that changes if we begin to use this estimate to filter out subnets entirely.

## What are the issues with the proposed solution?

Because the tracked IP count is not a lower-bound, the change made in this PR will not prevent cases where pods fail to start due to insufficient IPs. If Karpenter overestimates the number of IPs available in a subnet, the same issue will continue to occur. The current change may mitigate the issue in some cases, but due to the way we currently resync IP count from EC2 it will only be effective if nodes are consistently at max capacity.

Additionally, this change would be overly pessimistic for the majority of users. If node density is dictated by something other than maxPods (e.g. CPU or memory resources), this is going to overestimate the number of IPs that could be in use in a subnet. If this change was enabled for all users, this would result in scaling failures when they would have otherwise succeeded. This is the reason we resync IPs the way we do - it reflects the true IP consumption in the subnet.

## What could we do in Karpenter to solve this issue?

If we ensure Karpenter's IP estimate is a lower-bound for the number of IPs available in a cluster, we should be able to prevent this issue altogether. By ensuring that the estimate is a lower-bound, we ensure that no combination of pods scheduled to nodes in the subnet would exceed the number of IPs available in the subnets. However, this would be **far** too pessimistic for most workloads and is not a behavior we would want to support in Karpenter - it's a massive footgun. Without carefully tuning `maxPods` values on a per-instance basis to reflect real pod density, this would result in scaling failures which would not have occurred otherwise. For this reason, we would not want to go forward with this solution.

## What can we do? 

We acknowledge that this is an issue Karpenter users face, but we don't believe there is an adequate solution in Karpenter alone. There's two reasons for this:
- Karpenter is not the component responsible for binding pods to nodes - kube-scheduler is
- Karpenter is not a good source of truth for IP availability. The EC2 APIs are eventually consistent and we will never have anything better than a prediction.

To solve this issue, I believe we need a solution which involves both the CNI and kube-scheduler. In an ideal world, the scheduler would not bind a pod to a node without the CNI first allocating an IP for that pod. Alternatively, we could have a construct for resource limits within a topology (and introduce a subnet topology key). I imagine either of these proposals would need to take the form of a KEP. This is an issue we would like to tackle, but don't have timelines to share at the moment wrt when it would be prioritized.

There are also architectural changes that can be considered which wouldn't require changes to Karpenter or upstream Kubernetes:
- Using [custom networking](https://docs.aws.amazon.com/eks/latest/userguide/cni-custom-network.html) to place pods in a separate, less constrained subnet
- Tuning CNI configuration options to reduce unnecessary IP use (e.g. `WARM_IP_TARGET`, `WARM_ENI_TARGET`, and `MINIMUM_IP_TARGET`)
- Single subnet nodepools with pod limits

## Next Steps

As I mentioned earlier, we do understand this is an issue facing some Karpenter users and architectural improvements aren't always an option. We do plan on working on this issue, but our focus will be addressing it upstream rather than implementing a mitigation within Karpenter.

While I don't see a great path for the approach in this PR, we would still review an RFC that addresses the short-comings and sharp-edges discussed above. Opening an RFC does not guarantee that we would move forward with the change - it would still need to be reviewed and approved - but it is the potential path forward for a mitigation within Karpenter.
